docs/canvas/canvas-expression-lifecycle.asciidoc

```diff
@@ -0,0 +1,241 @@
+
+[[canvas-expressions-composing-functions-with-subexpressions]]
+=== Composing Functions with Sub-Expressions
+
+You may have noticed another syntax in examples from other sections, namely expressions inside of curly brackets. These are called sub-expressions, and they can be used to provide a calculated value to another expression, instead of just a static one.
+
+A simple example of this is when you upload your own images to a Canvas workpad. That upload becomes an asset, and that asset can be retrieved using the `asset` function. Usually you’ll just do this from the UI, adding an image element to the page and uploading your image from the control in the sidebar, or picking an existing asset from there as well. In both cases, the system will consume that asset via the `asset` function, and you’ll end up with an expression similar to this:
+
+*[NOTE] The examples in this section can’t be copied and pasted directly, since the values used throughout will not exist in your workpad*
+
+[source,text]
+----
+image dataurl={asset 3cb3ec3a-84d7-48fa-8709-274ad5cc9e0b}
+----
+
+Sub-expressions are executed before the function that uses them is executed. In this case, `asset` will be run first, it will produce a value, the base64-encoded value of the image, and that value will be used as the value for the `dataurl` argument in the <<image_fn, image>> function. So at first pass, you’ll have something like this, after the asset function executes:
+
+[source,text]
+----
+image dataurl="data:image/svg+xml;base64,PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0"
+----
+
+Since all of the sub-expressions are now resolved into actual values, the <<image_fn, image>> function can be executed to produce its JSON output, just as it’s explained previously. In the case of images, the ability to nest sub-expressions is particularly useful to show one of several images conditionally. For example, you could swap between two images based on some calculated value by mixing in the <<if_fn, if>> function, like in this example expression:
+
+[source,text]
+----
+demodata
+| image dataurl={
+  if condition={getCell price | gte 100} 
+    then={asset 3cb3ec3a-84d7-48fa-8709-274ad5cc9e0b} 
+    else={asset cbc11a1f-8f25-4163-94b4-2c3a060192e7}
+}
+----
+
+Here, the expression to use for the value of the `condition` argument, `getCell price | gte 100`, runs first since it is nested deeper. This expressions retrieves the value from the “price” column in the first row of the datatable produced by `demodata` and piping the value to the `gte` function, comparing the value to `100` and returning `true` if it’s 100 or more and `false` if it’s under 100. That boolean value becomes the value for the `condition` argument and the output from the `then` expression is used as the output when `condition` is `true`, while the output from the "else" expression is used when `condition` is false. In both cases, a base64-encoded image will be returned and one of the two images will be displayed.
```

Comment:
Since the second sentence is three lines long, how about breaking it up. For example:

The expression does the following:
* Retrieves the value from the **price** column in the first row of the `demodata` data table
* Inputs the value to the `gte` function
* Compares the data to `100`
* Returns `true` if the value is 100 or greater, or `false` if the value is 100 or less